pkg/user/tx_client_test.go

```diff
@@ -208,11 +245,62 @@ func (suite *TxClientTestSuite) TestConfirmTx() {
 		msg := bank.NewMsgSend(addr, testnode.RandomAddress().(sdk.AccAddress), sdk.NewCoins(sdk.NewInt64Coin(app.BondDenom, 1+balance)))
 		resp, err := suite.txClient.BroadcastTx(suite.ctx.GoContext(), []sdk.Msg{msg}, fee, gas)
 		require.NoError(t, err)
-		require.NotNil(t, resp)
-		_, err = suite.txClient.ConfirmTx(suite.ctx.GoContext(), resp.TxHash)
+		require.Equal(t, resp.Code, abci.CodeTypeOK)
+
+		nonce, signer, exists := suite.txClient.GetTxFromLocalMempool(resp.TxHash)
+		require.True(t, exists)
+		require.Equal(t, suite.txClient.DefaultAccountName(), signer)
+		seq := suite.txClient.Signer().Account(suite.txClient.DefaultAccountName()).Sequence()
+		// Successfully broadcast transaction increases the nonce
+		// txInfo is indexed before the nonce is increased
+		require.Equal(t, seq, nonce+1)
+
+		confirmTxResp, err := suite.txClient.ConfirmTx(suite.ctx.GoContext(), resp.TxHash)
 		require.Error(t, err)
+		require.Nil(t, confirmTxResp)
 		code := err.(*user.ExecutionError).Code
 		require.NotEqual(t, abci.CodeTypeOK, code)
+		nonce, signer, exists = suite.txClient.GetTxFromLocalMempool(resp.TxHash)
+		require.False(t, exists)
+		require.Zero(t, nonce)
+		require.Zero(t, signer)
+	})
+
+	t.Run("should adjust nonce when evicted", func(t *testing.T) {
+		// Set ttl to 1ns to evict tx immediately
+		suite.testnodeConfig.TmConfig.Mempool.TTLDuration = 1 * time.Nanosecond
+
+		// Keep submitting the transaction until we get the eviction error
+		sender := suite.txClient.Signer().Account(suite.txClient.DefaultAccountName())
+		msg := bank.NewMsgSend(sender.Address(), testnode.RandomAddress().(sdk.AccAddress), sdk.NewCoins(sdk.NewInt64Coin(app.BondDenom, 10)))
+		var seqBeforeEviction uint64
+		for {
+			// There is a chance one of the txs will get into a block so we have to keep updating the sequence
+			seqBeforeEviction = sender.Sequence()
+			resp, err := suite.txClient.BroadcastTx(suite.ctx.GoContext(), []sdk.Msg{msg}, fee, gas)
+			require.NoError(t, err)
+			_, err = suite.txClient.ConfirmTx(suite.ctx.GoContext(), resp.TxHash)
+			if err != nil {
+				if err.Error() == "tx was evicted from the mempool" {
+					break
+				}
+			}
+		}
+
+		seqAfterEviction := sender.Sequence()
+		require.Equal(t, seqBeforeEviction, seqAfterEviction)
+
+		// Increase ttl again
+		suite.testnodeConfig.TmConfig.Mempool.TTLDuration = 120 * time.Second
```

Comment:
do we need to restart the node for it to take effect?

Reply:
no because it's a pointer in the config and mempool just reads the config so if you update smth in the config it will automatically reference this new value

Reply:
would this then be the source a data race in our tests?

Reply:
I don't think so because in this test we're not sending so many txs plus other tests using tesnode would initialise the network with the default config overriding this ttl. I'd be happy to reset this back to the default though.

Reply:
it seems like we're writing to some state that is continually being read by the mempool, and there are no guarantees that we are not getting a data race, which is less than ideal. It's only in a tests and not prod, but I think the only reason we're not seeing it atm is because we aren't running this test with the race detector

the only way around this is to not change the config after it is passed to testnode, or we have to modify the usage in comet to first check a mutex before accessing the state

Reply:
Yeah I think if we want this to be done safely, it should be a separate test and we should only set it at the start

Reply:
yeah if i decrease the ttl at initialisation do you suggest that I reinitialise the network once I want to submit another tx which requires me to increase the ttl again?